Unable to search for tags when they contain a `'` in the tag value

This is a bug that seems to affects tags that contain a ' in the tag value. For example, this tag `bfkjefw'pov` cannot be searched via sql.

The bug seems to be caused by an issue in the underlying sql query and probably has an easy fix.

<img width="974" alt="image" src="https://user-images.githubusercontent.com/40373468/93693777-afdd7f00-fb47-11ea-8a9c-32e187d5b492.png">

bug details:

```
Caused by: org.postgresql.util.PSQLException: Unterminated identifier started at position 118 in SQL SELECT * FROM repo r WHERE r.user_id = '94866216-2343-4e1c-bf58-a30d0f941aec' and r.metadata @> '{"tags":["bfkjefw'pov"]}' order by r.repo_name asc. Expected " char
```

--- a/backend/build.gradle.kts
+++ b/backend/build.gradle.kts
@@ -31,14 +31,15 @@ dependencies {
     // utils
     implementation("com.vladmihalcea:hibernate-types-55:2.14.0")
     implementation("com.fasterxml.jackson.module:jackson-module-kotlin")
-    implementation("org.flywaydb:flyway-core:8.3.0")
+    implementation("org.flywaydb:flyway-core:8.5.7")
     implementation("io.github.microutils:kotlin-logging:2.1.21")
     implementation("io.jsonwebtoken:jjwt-api:0.11.2")
     implementation("io.jsonwebtoken:jjwt-jackson:0.11.2")
+    implementation("org.junit.jupiter:junit-jupiter:5.8.2")
     runtimeOnly("io.jsonwebtoken:jjwt-impl:0.11.2")
     implementation(platform("org.testcontainers:testcontainers-bom:1.14.3")) //bom for testcontainers
     // see: https://github.com/netty/netty/issues/11693
-    runtimeOnly("io.netty:netty-resolver-dns-native-macos:4.1.72.Final:osx-aarch_64")
+    runtimeOnly("io.netty:netty-resolver-dns-native-macos:4.1.75.Final:osx-aarch_64")
 
     // kotlin
     implementation("io.projectreactor.kotlin:reactor-kotlin-extensions")
@@ -55,12 +56,12 @@ dependencies {
     testImplementation("org.testcontainers:testcontainers")
     testImplementation("org.testcontainers:postgresql")
     testImplementation("org.testcontainers:junit-jupiter")
-    testImplementation("io.mockk:mockk:1.12.1")
+    testImplementation("io.mockk:mockk:1.12.3")
     testImplementation("com.willowtreeapps.assertk:assertk-jvm:0.25")
     testImplementation("io.projectreactor:reactor-test")
-    testImplementation("io.github.serpro69:kotlin-faker:1.9.0")
+    testImplementation("io.github.serpro69:kotlin-faker:1.10.0")
     testImplementation("org.apache.commons:commons-math3:3.6.1")
-    testImplementation("io.kotest:kotest-assertions-core:5.0.3")
+    testImplementation("io.kotest:kotest-assertions-core:5.2.2")
 }
 
 tasks.withType<Test> {

--- a/backend/src/main/kotlin/io/shiveenp/taggit/db/Repositories.kt
+++ b/backend/src/main/kotlin/io/shiveenp/taggit/db/Repositories.kt
@@ -6,9 +6,9 @@ import org.springframework.stereotype.Repository
 import java.util.*
 
 @Repository
-interface TaggitUserRepository : CrudRepository<TaggitUserEntity, UUID> {
+interface UserRepository : CrudRepository<TaggitUserEntity, UUID> {
     fun findByGithubUserId(githubUserId: Long): TaggitUserEntity?
 }
 
 @Repository
-interface TaggitRepoRepository : PagingAndSortingRepository<TaggitRepoEntity, UUID>
+interface RepoRepository : PagingAndSortingRepository<TaggitRepoEntity, UUID>

--- a/backend/src/main/kotlin/io/shiveenp/taggit/service/RepoSyncService.kt
+++ b/backend/src/main/kotlin/io/shiveenp/taggit/service/RepoSyncService.kt
@@ -1,7 +1,7 @@
 package io.shiveenp.taggit.service
 
 import io.shiveenp.taggit.db.TaggitRepoEntity
-import io.shiveenp.taggit.db.TaggitRepoRepository
+import io.shiveenp.taggit.db.RepoRepository
 import io.shiveenp.taggit.models.GithubStargazingResponse
 import io.shiveenp.taggit.util.notContains
 import mu.KotlinLogging
@@ -16,7 +16,7 @@ import org.springframework.stereotype.Service
 @Service
 class RepoSyncService(
     private val githubService: GithubService,
-    private val taggitRepoRepository: TaggitRepoRepository
+    private val repoRepository: RepoRepository
 ) {
 
     private val logger = KotlinLogging.logger { }
@@ -24,12 +24,12 @@ class RepoSyncService(
     fun syncUserStargazingData() {
         logger.debug { "Initiating repo sync..." }
         val repos = getUserStarredReposToSync()
-        val existingRepos = taggitRepoRepository.findAll()
+        val existingRepos = repoRepository.findAll()
         val currentSyncedRepoIds = existingRepos.map { it.repoId }
         repos.filter {
             currentSyncedRepoIds.notContains(it.id)
         }.forEach {
-            taggitRepoRepository.save(TaggitRepoEntity.from(it))
+            repoRepository.save(TaggitRepoEntity.from(it))
         }
         logger.debug { "Finished syncing repos!" }
     }

--- a/backend/src/main/kotlin/io/shiveenp/taggit/service/TaggitService.kt
+++ b/backend/src/main/kotlin/io/shiveenp/taggit/service/TaggitService.kt
@@ -23,8 +23,8 @@ import javax.persistence.EntityManagerFactory
 @Service
 class TaggitService(
     private val githubService: GithubService,
-    private val userRepository: TaggitUserRepository,
-    private val repoRepository: TaggitRepoRepository,
+    private val userRepository: UserRepository,
+    private val repoRepository: RepoRepository,
     private val entityManagerFactory: EntityManagerFactory,
     private val requestQueueRepository: RequestQueueRepository,
     private val mapper: ObjectMapper
@@ -102,10 +102,18 @@ class TaggitService(
         if (isUntaggedKeyword) {
             throw IllegalArgumentException("Cannot use 'untagged' as it's reserved")
         }
+        val sanitizedTag = sanitizeRepoTag(tagInput.tag)
+        if (sanitizedTag.isBlank()) {
+            throw IllegalArgumentException("The tag does not contain acceptable characters")
+        }
         return repoRepository.findByIdOrNull(repoId)?.let {
-            val updatedMetadata = addTagToMetadata(it.metadata, tagInput.tag)
+            val updatedMetadata = addTagToMetadata(it.metadata, sanitizedTag)
             repoRepository.save(it.withUpdated(metadata = updatedMetadata))
         }
+    }
+
+    private fun sanitizeRepoTag(repoTag: String): String {
+        return repoTag.filter { it.isLetterOrDigit() }
     }
 
     private fun addTagToMetadata(metadata: TagMetadata?, tag: String): TagMetadata {

--- a/backend/src/test/kotlin/io/shiveenp/taggit/integration/RepoTagControllerTest.kt
+++ b/backend/src/test/kotlin/io/shiveenp/taggit/integration/RepoTagControllerTest.kt
@@ -1,7 +1,11 @@
-package io.shiveenp.taggit
+package io.shiveenp.taggit.integration
 
-import io.shiveenp.taggit.db.TaggitRepoRepository
-import io.shiveenp.taggit.db.TaggitUserRepository
+import io.shiveenp.taggit.TaggitApplication
+import io.shiveenp.taggit.db.RepoRepository
+import io.shiveenp.taggit.db.UserRepository
+import io.shiveenp.taggit.generateMockRepoEntity
+import io.shiveenp.taggit.generateMockUserEntity
+import io.shiveenp.taggit.generateRandomTagInput
 import io.shiveenp.taggit.models.TagMetadata
 import org.hamcrest.Matchers
 import org.hamcrest.Matchers.hasItem
@@ -35,10 +39,10 @@ class RepoTagControllerTest {
     private val testUser = generateMockUserEntity()
 
     @Autowired
-    private lateinit var taggitUserRepository: TaggitUserRepository
+    private lateinit var userRepository: UserRepository
 
     @Autowired
-    private lateinit var taggitRepoRepository: TaggitRepoRepository
+    private lateinit var repoRepository: RepoRepository
 
     companion object {
 
@@ -63,7 +67,7 @@ class RepoTagControllerTest {
     @Test
     fun `we can save single tags correctly`() {
         val testRepo = generateMockRepoEntity(testUser.id)
-        taggitRepoRepository.save(testRepo)
+        repoRepository.save(testRepo)
         val tagInputToSave = generateRandomTagInput()
         webTestClient.post()
             .uri("/user/${testUser.id}/repos/${testRepo.id}/tag")
@@ -81,7 +85,7 @@ class RepoTagControllerTest {
     @Test
     fun `we can save multiple tags correctly`() {
         val testRepo = generateMockRepoEntity(testUser.id)
-        taggitRepoRepository.save(testRepo)
+        repoRepository.save(testRepo)
         val tagInputToSave1 = generateRandomTagInput()
         val tagInputToSave2 = generateRandomTagInput()
         // save first tag
@@ -120,7 +124,7 @@ class RepoTagControllerTest {
     fun `we can delete single tags correctly`() {
         val tagToDelete = "hello,fake-tag"
         val testRepo = generateMockRepoEntity(testUser.id, TagMetadata(listOf(tagToDelete)))
-        taggitRepoRepository.save(testRepo)
+        repoRepository.save(testRepo)
 
         // technically you wanna check that the tag was saved properly before checking again,
         // but in this case we're trusting the repository impl and would hopefully test that in
@@ -143,7 +147,7 @@ class RepoTagControllerTest {
         val tagToDelete3 = "hello, fake tag 3"
         val testRepo = generateMockRepoEntity(testUser.id,
             TagMetadata(listOf(tagToDelete1, tagToDelete2, tagToDelete3)))
-        taggitRepoRepository.save(testRepo)
+        repoRepository.save(testRepo)
 
         // 1 deleted, 2 remaining
         webTestClient.delete()
@@ -194,7 +198,7 @@ class RepoTagControllerTest {
         val testRepo = generateMockRepoEntity(testUser.id,
             TagMetadata(listOf(tagToDelete1, tagToDelete2, tagToDelete3))
         )
-        taggitRepoRepository.save(testRepo)
+        repoRepository.save(testRepo)
 
         // 1 deleted, 3 remaining
         webTestClient.delete()
@@ -239,6 +243,6 @@ class RepoTagControllerTest {
 
     @BeforeAll
     internal fun setUp() {
-        taggitUserRepository.save(testUser)
+        userRepository.save(testUser)
     }
 }

--- a/backend/src/test/kotlin/io/shiveenp/taggit/service/TaggitServiceTest.kt
+++ b/backend/src/test/kotlin/io/shiveenp/taggit/service/TaggitServiceTest.kt
@@ -1,0 +1,95 @@
+package io.shiveenp.taggit.service
+
+import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
+import io.kotest.assertions.throwables.shouldThrow
+import io.kotest.matchers.shouldBe
+import io.kotest.matchers.shouldNotBe
+import io.mockk.every
+import io.mockk.impl.annotations.MockK
+import io.mockk.impl.annotations.OverrideMockKs
+import io.mockk.junit5.MockKExtension
+import io.shiveenp.taggit.db.RepoRepository
+import io.shiveenp.taggit.db.RequestQueueRepository
+import io.shiveenp.taggit.db.UserRepository
+import io.shiveenp.taggit.generateMockRepoEntity
+import io.shiveenp.taggit.models.TagInput
+import kotlinx.coroutines.runBlocking
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.extension.ExtendWith
+import org.springframework.data.repository.findByIdOrNull
+import java.util.*
+import javax.persistence.EntityManagerFactory
+
+@ExtendWith(MockKExtension::class)
+class TaggitServiceTest {
+
+    @MockK
+    private lateinit var githubService: GithubService
+    @MockK
+    private lateinit var userRepository: UserRepository
+    @MockK
+    private lateinit var repoRepository: RepoRepository
+    @MockK
+    private lateinit var entityManagerFactory: EntityManagerFactory
+    @MockK
+    private lateinit var requestQueueRepository: RequestQueueRepository
+
+    private val repoEntity = generateMockRepoEntity()
+    private val mapper = jacksonObjectMapper()
+
+    @OverrideMockKs
+    private lateinit var taggitService: TaggitService
+
+    @BeforeEach
+    fun setup() {
+        every { repoRepository.findByIdOrNull(any()) } returns repoEntity
+        every { repoRepository.save(any()) } returnsArgument 0
+    }
+
+    @Test
+    fun `add repo tag returns the right data on successfully save`() {
+        val repoId = UUID.randomUUID()
+        val tagInput = TagInput("test")
+        runBlocking {
+            val repoEntity = taggitService.addRepoTag(repoId, tagInput)
+            repoEntity shouldNotBe null
+            repoEntity?.metadata shouldNotBe null
+            repoEntity!!.metadata!!.tags shouldBe listOf("test")
+        }
+    }
+
+    @Test
+    fun `add repo tag throws exception if user tries to save untagged keyword`() {
+        val repoId = UUID.randomUUID()
+        val tagInput = TagInput("untagged")
+        shouldThrow<IllegalArgumentException> {
+            runBlocking {
+                taggitService.addRepoTag(repoId, tagInput)
+            }
+        }
+    }
+
+    @Test
+    fun `should strip non alphanumeric characters from the tag string`() {
+        val repoId = UUID.randomUUID()
+        val tagInput = TagInput("test-new'")
+        runBlocking {
+            val repoEntity = taggitService.addRepoTag(repoId, tagInput)
+            repoEntity shouldNotBe null
+            repoEntity?.metadata shouldNotBe null
+            repoEntity!!.metadata!!.tags shouldBe listOf("testnew")
+        }
+    }
+
+    @Test
+    fun `should throw exception if blank string left after stripping non alphanumeric chars`() {
+        val repoId = UUID.randomUUID()
+        val tagInput = TagInput("'#$%%^@")
+        shouldThrow<IllegalArgumentException> {
+            runBlocking {
+                taggitService.addRepoTag(repoId, tagInput)
+            }
+        }
+    }
+}

--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -3,7 +3,7 @@ services:
   taggit:
     profiles: ["prod"]
     container_name: 'taggit'
-    image: 'shiveenp/taggit:latest'
+    image: 'ghcr.io/shiveenp/shiveenp/taggit:latest'
     ports:
       - '127.0.0.1:8080:8080'
     depends_on:

--- a/frontend/src/components/GitHubRepo.vue
+++ b/frontend/src/components/GitHubRepo.vue
@@ -48,12 +48,13 @@ export default {
   name: "GithubRepo",
   methods: {
     saveTag(tag) {
+      let sanitizedTag = this.sanitizeTag(tag);
       this.$store.dispatch('addTagToRepo',
           {
             repoId: this.id,
-            tag: tag
+            tag: sanitizedTag
           });
-      this.$store.dispatch('addTag', tag);
+      this.$store.dispatch('addTag', sanitizedTag);
     },
     removeTag(tag) {
       axios.delete('/api/repos/' + this.id + '/tag?tag=' + tag);
@@ -66,6 +67,9 @@ export default {
     },
     makeTagInputVisible() {
       this.isTagInputVisible = !this.isTagInputVisible
+    },
+    sanitizeTag(tag) {
+      return tag.replace(/[^a-zA-Z0-9]/g, '')
     }
   },
   mounted() {

